### PR TITLE
More clever generaton of the factory_attributes

### PR DESCRIPTION
**Pull request by dima4p**
For complicated models this patch drastically reduces the developers work.

### CODE REVIEW

File: lib/generators/factory_bot/model/model_generator.rb [hunk 1]
```diff
@@ -62,8 +62,26 @@ def factory_definition
...
+          elsif attribute.name == 'email'
+            "sequence(:#{attribute.name}) {|n| \"email#\{format '%03d', n}@gmail.com\" }"
+          elsif attribute.name =~ /(.*)_url$/
+            "sequence(:#{attribute.name}) {|n| \"http://#\{$1}#\{format '%03d', n}.com\" }"
+          elsif attribute.name == 'password'
+            "password 'password'"
+          elsif attribute.name == 'position'
+            "sequence(:#{attribute.name}) {|n| n }"
+          elsif [:string, :text].include? attribute.type
+            "sequence(:#{attribute.name}) {|n| \"#{attribute.name.capitalize.gsub('_', ' ')}#\{format '%03d', n}\" }"
```
[the remark this answers is not in the document]
Comment (dima4p) [reply lib/generators/factory_bot/model/model_generator.rb]: Should I divide a string into the two and sum them?

**Comment by dima4p** [comment]
Sorry, I do not know how to satisfy codeclimate.

**Comment by dima4p** [comment]
I need hints how to fix.

**Comment by composerinteralia** [comment]
Thanks for the PR!

@dima4p codeclimate is complaining about the cognitive complexity of the method, since this is one long method with lots of conditionals.

I like some of the changes here (like using an association when the attribute is a reference), but some of the others make assumptions about column names that I am not sure will hold up in every case.

Could we start with a smaller PR including just the association change, since I think that change will be the easiest to merge? That PR would need to include test coverage for the change.

**Comment by dima4p** [comment]
> Thanks for the PR!
> 
> @dima4p codeclimate is complaining about the cognitive complexity of the method, since this is one long method with lots of conditionals.

I see that. The problem is the decomposition will make the code to be better for robot but worse for human.
> 
> I like some of the changes here (like using an association when the attribute is a reference), but some of the others make assumptions about column names that I am not sure will hold up in every case.
> 
> Could we start with a smaller PR including just the association change, since I think that change will be the easiest to merge? That PR would need to include test coverage for the change.

Indeed all the cases were defined many years ago and used in multiply projects. They are all required. Those time there was factory_girl gem for which I have created this [gem](https://github.com/dima4p/factory_girl_fixtures_template).

And this PR is just the transition of it.

**Comment by composerinteralia** [comment]
> I see that. The problem is the decomposition will make the code to be better for robot but worse for human.

I think I am in agreement with code climate on this one. I find this method rather difficult to read. Smaller methods with clear names are generally easier than one long method, so you could start refactoring by pulling out smaller helper methods.

> Indeed all the cases were defined many years ago and used in multiply projects. They are all required. Those time there was factory_girl gem for which I have created this gem.

These may be the attributes required for your own projects, but they are not necessarily the attributes required for every project. Smaller PRs would be helpful here because it would allow us to discuss each attribute separately and decide whether it is a general use case that belongs in factory_bot_rails.

**Comment by composerinteralia** [comment]
Since there is no movement on this, I am going to close for now. I do think this is an interesting idea and would be happy to explore it in a series of smaller PRs.

**Comment by dima4p** [comment]
Is soon as I have found some time for doing this in pieces, I will do it. Up to now all are welcome to my fork.

**Pull request closed without merging**